File: peek_plugin_graphdb/_private/tuples/ItemKeyTuple.py
import json
from typing import Dict

from peek_plugin_graphdb._private.PluginNames import graphDbTuplePrefix
from peek_plugin_graphdb._private.tuples.ItemKeyTypeTuple import ItemKeyTypeTuple
from vortex.Tuple import Tuple, addTupleType, TupleField


class GraphDbModelSetTuple(object):
    pass


@addTupleType
class ItemKeyTuple(Tuple):
    """ ItemKeyIndex Tuple

    This tuple is the publicly exposed ItemKeyIndex

    """
    __tupleType__ = graphDbTuplePrefix + 'ItemKeyTuple'

    #:  The unique key of this itemKeyIndex
    key: str = TupleField()

    #:  The model set of this itemKeyIndex
    modelSet: GraphDbModelSetTuple = TupleField()

    #:  The itemKeyIndex type
    itemKeyType: ItemKeyTypeTuple = TupleField()

    #:  A string value of the itemKey
    valueStr: Dict = TupleField()

    #:  An int value of the itemKey
    valueInt: Dict = TupleField()

    # Add more values here

    @classmethod
    def unpackJson(self, key: str, packedJson: str):
        # Reconstruct the data
        objectProps: {} = json.loads(packedJson)

        # Get out the object type
        thisItemKeyTypeId = objectProps['_tid']

        # Get out the object type
        thisModelSetId = objectProps['_msid']

        # Create the new object
        newSelf = ItemKeyTuple()

        newSelf.key = key

        # These objects get replaced with the full object in the UI
        newSelf.modelSet = GraphDbModelSetTuple(id__=thisModelSetId)
        newSelf.itemKeyType = ItemKeyTypeTuple(id__=thisItemKeyTypeId)

        # Unpack the custom data here
        newSelf.valueStr = objectProps.get('valueStr')
        newSelf.valueInt = objectProps.get('valueInt')

        return newSelf

File: peek_plugin_graphdb/plugin-module/_private/item-key-index-loader/ItemKeyTuple.ts
import {addTupleType, Tuple} from "@synerty/vortexjs";
import {graphDbTuplePrefix} from "../PluginNames";
import {GraphDbModelSetTuple} from "../../GraphDbModelSetTuple";


@addTupleType
export class ItemKeyTuple extends Tuple {
    public static readonly tupleName = graphDbTuplePrefix + "ItemKeyTuple";

    //  The modelSetId for this itemKeyIndex.
    modelSet: GraphDbModelSetTuple = new GraphDbModelSetTuple();

    // The key of the vertex or edge
    itemKey: string;

    // This ItemKeyIndex Type ID
    itemType: number;
    static readonly ITEM_TYPE_VERTEX = 1;
    static readonly ITEM_TYPE_EDGE = 2;

    // The key of the segment where it's stored
    segmentKeys: string[];

    constructor() {
        super(ItemKeyTuple.tupleName)
    }

    static unpackJson(key: string, packedJson: string): ItemKeyTuple {
        // Reconstruct the data
        let objectProps: {} = JSON.parse(packedJson);

        // Get out the object type
        let thisModelSetId = objectProps['_msid'];
        delete objectProps['_msid'];

        // Create the new object
        let newSelf = new ItemKeyTuple();

        // These objects get replaced later in the UI
        newSelf.modelSet = new GraphDbModelSetTuple();
        newSelf.modelSet.id = thisModelSetId;

        // Unpack the custom data here
        newSelf.itemKey = key;
        newSelf.itemType = objectProps["itemType"];
        newSelf.segmentKeys = objectProps["segmentKeys"];

        return newSelf;

    }
}